appserver/admin/cli-optional/src/main/java/com/sun/enterprise/admin/cli/optional/DerbyExecuteSQL.java

```diff
@@ -0,0 +1,104 @@
+
+import java.io.File;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.util.Properties;
+import java.util.Scanner;
+
+import static com.sun.enterprise.util.Utility.isAllEmpty;
+
+/**
+ * This class executes SQL statements from a file against Derby. Its prime purpose is for database initialization
+ * after starting it.
+ *
+ * <p>
+ * The reason for creating this class instead of directly executing SQL from the
+ * StartDatabaseCommand class is so that a separate JVM is launched when starting the database with all the right
+ * drivers in place for the remote Derby.
+ *
+ * @author Arjan Tijms
+ */
+public class DerbyExecuteSQL {
+
+    final private String derbyHost;
+    final private String derbyPort;
+    final private String derbyName;
+    final private String sqlFileName;
+    final private String derbyUser;
+    final private String derbyPassword;
+
+    public static void main(String[] args) {
+        DerbyExecuteSQL sql = new DerbyExecuteSQL(
+                args[0],
+                args[1],
+                args[2],
+                args[3],
+                args.length < 5? null : args[4],
+                args.length < 6? null : args[5]);
+
+        sql.executeCommand();
+    }
+
+    public DerbyExecuteSQL(String derbyHost, String derbyPort, String derbyName, String sqlFileName, String derbyUser, String derbyPassword) {
+        this.derbyHost = derbyHost;
+        this.derbyPort = derbyPort;
+        this.derbyName = derbyName;
+        this.sqlFileName = sqlFileName;
+        this.derbyUser = derbyUser;
+        this.derbyPassword = derbyPassword;
+    }
+
+    private void executeCommand() {
+        try {
+            Driver driver = (Driver)
+                Class.forName("org.apache.derby.jdbc.ClientDriver")
+                     .getDeclaredConstructor()
+                     .newInstance();
+
+            Properties properties = new Properties();
+            if (!isAllEmpty(derbyUser, derbyPassword)) {
+                properties.put("user", derbyUser);
+                properties.put("password", derbyPassword);
+            }
+
+            try (Connection connection = driver.connect("jdbc:derby://" + derbyHost + ":" + derbyPort + "/" + derbyName, new Properties())) {
+                try (Scanner scanner = new Scanner(new File(sqlFileName), "UTF-8")) {
+
+                    scanner.useDelimiter(";");
+
+                    while (scanner.hasNext()) {
+                        try {
+                            String statement = scanner.next().strip();
+                            if (!statement.isEmpty()) {
+                                connection.prepareStatement(statement).execute();
+                            }
+                        } catch (SQLException s) {
+                            s.printStackTrace();
+                        }
+                    }
+                }
+            }
+        } catch (Throwable e) {
+            e.printStackTrace();
```

Comment:
Wouldn't be better to throw it out? It should crash when it has broken script ...

Reply:
It should not crash, although we can add an option to crash or not.

Scripts very often do a "drop table", and don't have logic to do "drop table if not exist".

Reply:
The SQL ANT task (which the TCK often uses to init data) behaves the same way. I recently checked it out. It has an option to fail right away, and at the end it prints the number of failed / succeeded statements.

It has a ton of more options, but see: https://ant.apache.org/manual/Tasks/sql.html

Reply:
In this case is quite harmful to ignore the error instead of escalating it. Especially in TCK you simply don't want to search logs for a stacktrace when it should crash immediately. That is a good practice, if TCK doesn't follow that, we can fix it - I submitted fixes for JEE10 of this kind ;)
These hidden crashes took a lot of time.

Reply:
In the case of the TCK it's not a real error. Derby should use

"create if not exists table ..." or "drop if exists table..."

But that syntax is not there.

So it must do:

"drop table"
"create table ..."

And then ignore the "drop table" if it fails.